### ISSUE TITLE
confu.exceptions circular import

### ISSUE THREAD

**Issue by doteater** [20c/vaping#185]
Is this project abandoned? I've tried getting vaping to run on both Ubuntu 22.04 and current Arch, both running locally and using docker and every time I'm getting a circular import error.  Let me know if I'm doing something dumb or if I can provide any further info. 

### Arch + Docker
```
[user@archbox vaping]$ docker --version
Docker version 25.0.4, build 1a576c50a9
```
```
[user@archbox vaping]$ sudo pacman -Syu
:: Synchronizing package databases...
 core is up to date
 extra is up to date
:: Starting full system upgrade...
 there is nothing to do
```
```
[user@archbox ~]$ git clone https://github.com/20c/vaping
Cloning into 'vaping'...
remote: Enumerating objects: 2539, done.
remote: Counting objects: 100% (549/549), done.
remote: Compressing objects: 100% (208/208), done.
remote: Total 2539 (delta 368), reused 439 (delta 310), pack-reused 1990
Receiving objects: 100% (2539/2539), 3.69 MiB | 461.00 KiB/s, done.
Resolving deltas: 100% (1572/1572), done.
[user@archbox ~]$ cd vaping/
[user@archbox vaping]$ sudo docker build -t vaping .
DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
            Install the buildx component to build images with BuildKit:
            https://docs.docker.com/go/buildx/

Sending build context to Docker daemon  4.938MB
Step 1/51 : ARG build_packages="    gcc     g++     make     file     libc-dev     libffi-dev     curl     openssl-dev     make     rust     cargo     rrdtool-dev     "
Step 2/51 : ARG runtime_packages="fping librrd     zeromq     "
Step 3/51 : ARG vaping_extras=all
Step 4/51 : ARG poetry_pin=">=1,<=2"
Step 5/51 : FROM python:3.9-alpine as base
 ---> 196d79f7c614
Step 6/51 : ARG runtime_packages
 ---> Using cache
 ---> 597b3f1c0ba5
Step 7/51 : ARG virtual_env=/venv
 ---> Using cache
 ---> 60e9e4482500
Step 8/51 : ENV VIRTUAL_ENV="$virtual_env"
 ---> Using cache
 ---> a02cb595b930
Step 9/51 : ENV PATH="$VIRTUAL_ENV/bin:$PATH"
 ---> Using cache
 ---> 504305e4c0fd
Step 10/51 : FROM base as builder
 ---> 504305e4c0fd
Step 11/51 : ARG build_packages
 ---> Using cache
 ---> b2b1c156d832
Step 12/51 : ARG runtime_packages
 ---> Using cache
 ---> 4f61c29bc632
Step 13/51 : ARG vaping_extras
 ---> Using cache
 ---> ce98a12d99e8
Step 14/51 : RUN apk upgrade --available && apk --update --no-cache add $build_packages $runtime_packages
 ---> Using cache
 ---> 92916bc90141
Step 15/51 : RUN pip install --upgrade pip wheel
 ---> Using cache
 ---> a15fe33e8626
Step 16/51 : RUN pip install "poetry$poetry_pin"
 ---> Using cache
 ---> 3328b7562bc1
Step 17/51 : RUN python3 -m venv "$VIRTUAL_ENV"
 ---> Using cache
 ---> 33e2def0493d
Step 18/51 : WORKDIR /src/vaping
 ---> Using cache
 ---> 3981d7ee3155
Step 19/51 : COPY Ctl/VERSION Ctl/
 ---> Using cache
 ---> 2288af8644bd
Step 20/51 : COPY pyproject.toml poetry.lock README.md ./
 ---> Using cache
 ---> 8a25bcfa7175
Step 21/51 : COPY src src
 ---> Using cache
 ---> b4cfa3f8a599
Step 22/51 : RUN poetry run pip install --upgrade pip wheel
 ---> Using cache
 ---> efe5aed90307
Step 23/51 : RUN poetry build -f wheel
 ---> Using cache
 ---> 5d6a726240d7
Step 24/51 : RUN pip install "$(ls dist/vaping-*.whl)[${vaping_extras}]"
 ---> Using cache
 ---> 73ba1c10e510
Step 25/51 : FROM base as final
 ---> 504305e4c0fd
Step 26/51 : ARG runtime_packages
 ---> Using cache
 ---> e9a4af5a1c8b
Step 27/51 : ARG vaping_uid=1000
 ---> Using cache
 ---> 3c9c91cb6902
Step 28/51 : RUN apk upgrade --available     && apk --update --no-cache add $runtime_packages     && rm -rf /var/cache/apk/*
 ---> Using cache
 ---> d6446d7e4bb2
Step 29/51 : COPY --from=builder "$VIRTUAL_ENV" "$VIRTUAL_ENV"
 ---> Using cache
 ---> d31fa54524a7
Step 30/51 : RUN adduser -Du $vaping_uid vaping
 ---> Using cache
 ---> 7426cf27c332
Step 31/51 : FROM final as tester
 ---> 7426cf27c332
Step 32/51 : ARG build_packages
 ---> Using cache
 ---> f9d5eff56d48
Step 33/51 : ARG vaping_extras
 ---> Using cache
 ---> f20a92d19aa5
Step 34/51 : WORKDIR /src/vaping
 ---> Using cache
 ---> d853e0d26a7b
Step 35/51 : COPY pyproject.toml poetry.lock README.md ./
 ---> Using cache
 ---> ec40691326b0
Step 36/51 : COPY src src
 ---> Using cache
 ---> 30da1a9942a7
Step 37/51 : COPY tests tests
 ---> Using cache
 ---> 3a9ef826d412
Step 38/51 : RUN env
 ---> Using cache
 ---> 1b7f12301d77
Step 39/51 : RUN apk --update add $build_packages
 ---> Using cache
 ---> c3186b832ab8
Step 40/51 : RUN pip install "poetry$poetry_pin"
 ---> Using cache
 ---> 68fc8a2d0893
Step 41/51 : RUN poetry install --no-root --extras ${vaping_extras}
 ---> Using cache
 ---> cec36569318e
Step 42/51 : FROM final
 ---> 7426cf27c332
Step 43/51 : ARG vaping_home=/vaping
 ---> Running in bcb7084ddd07
 ---> Removed intermediate container bcb7084ddd07
 ---> cf9377d657b5
Step 44/51 : ENV VAPING_HOME=$vaping_home
 ---> Running in e03f83058662
 ---> Removed intermediate container e03f83058662
 ---> aa13141892d0
Step 45/51 : WORKDIR /vaping
 ---> Running in 1b11a3d578e8
 ---> Removed intermediate container 1b11a3d578e8
 ---> db1956d9d265
Step 46/51 : RUN chown vaping:vaping /vaping
 ---> Running in 2cc32b720239
 ---> Removed intermediate container 2cc32b720239
 ---> 677792afac30
Step 47/51 : USER vaping
 ---> Running in 9d3e53c71161
 ---> Removed intermediate container 9d3e53c71161
 ---> b2c389fcd0e0
Step 48/51 : COPY --chown=vaping:vaping examples examples
 ---> 5623c295c85a
Step 49/51 : COPY --chown=vaping:vaping examples/standalone_dns/config.yml .
 ---> 291339b71cdf
Step 50/51 : EXPOSE 7021
 ---> Running in 48bf4e710b75
 ---> Removed intermediate container 48bf4e710b75
 ---> 5c03daf6ef60
Step 51/51 : CMD ["vaping", "start", "--no-fork"]
 ---> Running in 004b110291c5
 ---> Removed intermediate container 004b110291c5
 ---> c072817b3389
Successfully built c072817b3389
Successfully tagged vaping:latest
[user@archbox vaping]$ sudo docker run vaping
Traceback (most recent call last):
  File "/venv/bin/vaping", line 5, in <module>
    from vaping.cli import cli
  File "/venv/lib/python3.9/site-packages/vaping/__init__.py", line 4, in <module>
    from vaping.config import Config  # noqa
  File "/venv/lib/python3.9/site-packages/vaping/config/__init__.py", line 5, in <module>
    from vaping.config.schema import *  # noqa
  File "/venv/lib/python3.9/site-packages/vaping/config/schema.py", line 1, in <module>
    import confu.exceptions
  File "/venv/lib/python3.9/site-packages/confu/exceptions.py", line 5, in <module>
    import confu.schema
  File "/venv/lib/python3.9/site-packages/confu/schema/__init__.py", line 1, in <module>
    from confu.schema.core import *  # noqa
  File "/venv/lib/python3.9/site-packages/confu/schema/core.py", line 18, in <module>
    from confu.exceptions import ApplyDefaultError, ValidationError, ValidationWarning
ImportError: cannot import name 'ApplyDefaultError' from partially initialized module 'confu.exceptions' (most likely due to a circular import) (/venv/lib/python3.9/site-packages/confu/exceptions.py)
```

### Arch w/o Docker 

```
[user@archbox ~]$ pipx install vaping
  installed package vaping 1.5.3, installed using Python 3.11.8
  These apps are now globally available
    - vaping
done! ✨ 🌟 ✨
[user@archbox ~]$ git clone https://github.com/20c/vaping
Cloning into 'vaping'...
remote: Enumerating objects: 2539, done.
remote: Counting objects: 100% (549/549), done.
remote: Compressing objects: 100% (208/208), done.
remote: Total 2539 (delta 368), reused 439 (delta 310), pack-reused 1990
Receiving objects: 100% (2539/2539), 3.69 MiB | 490.00 KiB/s, done.
Resolving deltas: 100% (1572/1572), done.
[user@archbox ~]$ cd vaping/
[user@archbox vaping]$ vaping start --home=examples/standalone_dns/ --debug
Traceback (most recent call last):
  File "/home/user/.local/bin/vaping", line 5, in <module>
    from vaping.cli import cli
  File "/home/user/.local/share/pipx/venvs/vaping/lib/python3.11/site-packages/vaping/__init__.py", line 4, in <module>
    from vaping.config import Config  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/share/pipx/venvs/vaping/lib/python3.11/site-packages/vaping/config/__init__.py", line 5, in <module>
    from vaping.config.schema import *  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/share/pipx/venvs/vaping/lib/python3.11/site-packages/vaping/config/schema.py", line 1, in <module>
    import confu.exceptions
  File "/home/user/.local/share/pipx/venvs/vaping/lib/python3.11/site-packages/confu/exceptions.py", line 5, in <module>
    import confu.schema
  File "/home/user/.local/share/pipx/venvs/vaping/lib/python3.11/site-packages/confu/schema/__init__.py", line 1, in <module>
    from confu.schema.core import *  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/share/pipx/venvs/vaping/lib/python3.11/site-packages/confu/schema/core.py", line 18, in <module>
    from confu.exceptions import ApplyDefaultError, ValidationError, ValidationWarning
ImportError: cannot import name 'ApplyDefaultError' from partially initialized module 'confu.exceptions' (most likely due to a circular import) (/home/user/.local/share/pipx/venvs/vaping/lib/python3.11/site-packages/confu/exceptions.py)
```

### Ubuntu + Docker

```
user@ubuntubox:~$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.4 LTS"
```
```
user@ubuntubox:~$ sudo docker --version
Docker version 25.0.4, build 1a576c5
```
```
user@ubuntubox:~$ git clone https://github.com/20c/vaping
Cloning into 'vaping'...
remote: Enumerating objects: 2539, done.
remote: Counting objects: 100% (549/549), done.
remote: Compressing objects: 100% (208/208), done.
remote: Total 2539 (delta 368), reused 439 (delta 310), pack-reused 1990
Receiving objects: 100% (2539/2539), 3.69 MiB | 420.00 KiB/s, done.
Resolving deltas: 100% (1572/1572), done.
user@ubuntubox:~$ cd vaping/
user@ubuntubox:~/vaping$ sudo docker build -t vaping .
[+] Building 320.7s (24/24) FINISHED                                                                                      docker:default
 => [internal] load build definition from Dockerfile                                                                                0.0s
 => => transferring dockerfile: 2.32kB                                                                                              0.0s
 => [internal] load metadata for docker.io/library/python:3.9-alpine                                                                0.3s
 => [internal] load .dockerignore                                                                                                   0.0s
 => => transferring context: 2B                                                                                                     0.0s
 => CACHED [base 1/1] FROM docker.io/library/python:3.9-alpine@sha256:0e9fe391e21bc8f2289934a9d0300c6801b4cf404340af735a0ca12abfd1  0.0s
 => [internal] load build context                                                                                                   0.0s
 => => transferring context: 2.52kB                                                                                                 0.0s
 => CACHED [final 1/3] RUN apk upgrade --available     && apk --update --no-cache add fping librrd     zeromq          && rm -rf /  0.0s
 => [builder  1/11] RUN apk upgrade --available && apk --update --no-cache add     gcc     g++     make     file     libc-dev     272.8s
 => [builder  2/11] RUN pip install --upgrade pip wheel                                                                             4.9s
 => [builder  3/11] RUN pip install "poetry$poetry_pin"                                                                            21.2s
 => [builder  4/11] RUN python3 -m venv "/venv"                                                                                     3.5s
 => [builder  5/11] WORKDIR /src/vaping                                                                                             0.1s
 => [builder  6/11] COPY Ctl/VERSION Ctl/                                                                                           0.0s
 => [builder  7/11] COPY pyproject.toml poetry.lock README.md ./                                                                    0.1s
 => [builder  8/11] COPY src src                                                                                                    0.1s
 => [builder  9/11] RUN poetry run pip install --upgrade pip wheel                                                                  3.1s
 => [builder 10/11] RUN poetry build -f wheel                                                                                       0.9s
 => [builder 11/11] RUN pip install "$(ls dist/vaping-*.whl)[all]"                                                                 11.7s
 => [final 2/3] COPY --from=builder /venv /venv                                                                                     0.3s
 => [final 3/3] RUN adduser -Du 1000 vaping                                                                                         0.3s
 => [stage-4 1/4] WORKDIR /vaping                                                                                                   0.1s
 => [stage-4 2/4] RUN chown vaping:vaping /vaping                                                                                   0.5s
 => [stage-4 3/4] COPY --chown=vaping:vaping examples examples                                                                      0.1s
 => [stage-4 4/4] COPY --chown=vaping:vaping examples/standalone_dns/config.yml .                                                   0.1s
 => exporting to image                                                                                                              0.3s
 => => exporting layers                                                                                                             0.3s
 => => writing image sha256:ee011d2b3b40b685072c3dc42954a42792a2706cd4212f7946c8287cd081aab2                                        0.0s
 => => naming to docker.io/library/vaping                                                                                           0.0s
user@ubuntubox:~/vaping$ sudo docker run vaping
Traceback (most recent call last):
  File "/venv/bin/vaping", line 5, in <module>
    from vaping.cli import cli
  File "/venv/lib/python3.9/site-packages/vaping/__init__.py", line 4, in <module>
    from vaping.config import Config  # noqa
  File "/venv/lib/python3.9/site-packages/vaping/config/__init__.py", line 5, in <module>
    from vaping.config.schema import *  # noqa
  File "/venv/lib/python3.9/site-packages/vaping/config/schema.py", line 1, in <module>
    import confu.exceptions
  File "/venv/lib/python3.9/site-packages/confu/exceptions.py", line 5, in <module>
    import confu.schema
  File "/venv/lib/python3.9/site-packages/confu/schema/__init__.py", line 1, in <module>
    from confu.schema.core import *  # noqa
  File "/venv/lib/python3.9/site-packages/confu/schema/core.py", line 18, in <module>
    from confu.exceptions import ApplyDefaultError, ValidationError, ValidationWarning
ImportError: cannot import name 'ApplyDefaultError' from partially initialized module 'confu.exceptions' (most likely due to a circular import) (/venv/lib/python3.9/site-packages/confu/exceptions.py)
```

### Ubuntu w/o Docker

```
user@ubuntubox:~$ pip install vaping
Defaulting to user installation because normal site-packages is not writeable
Collecting vaping
  Downloading vaping-1.5.3-py3-none-any.whl.metadata (6.0 kB)
Requirement already satisfied: Flask<2.2,>2 in /usr/local/lib/python3.10/dist-packages (from vaping) (2.1.3)
Requirement already satisfied: Werkzeug<2.1.0,>2 in /usr/local/lib/python3.10/dist-packages (from vaping) (2.0.3)
Requirement already satisfied: confu<2.0.0,>=1.7.1 in /usr/local/lib/python3.10/dist-packages (from vaping) (1.9.0)
Requirement already satisfied: munge<2.0.0,>=1.2.0 in /usr/local/lib/python3.10/dist-packages (from munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (1.3.0)
Requirement already satisfied: pid<4,>=3 in /usr/local/lib/python3.10/dist-packages (from vaping) (3.0.4)
Requirement already satisfied: pluginmgr<2,>=1 in /usr/local/lib/python3.10/dist-packages (from vaping) (1.1.0)
Requirement already satisfied: python-daemon<3,>=2 in /usr/local/lib/python3.10/dist-packages (from vaping) (2.3.2)
Requirement already satisfied: Jinja2>=3.0 in /usr/lib/python3/dist-packages (from Flask<2.2,>2->vaping) (3.0.3)
Requirement already satisfied: itsdangerous>=2.0 in /usr/local/lib/python3.10/dist-packages (from Flask<2.2,>2->vaping) (2.1.2)
Requirement already satisfied: click>=8.0 in ./.local/lib/python3.10/site-packages (from Flask<2.2,>2->vaping) (8.1.7)
Requirement already satisfied: requests>=2.6 in /usr/local/lib/python3.10/dist-packages (from munge<2.0.0,>=1.2.0->munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (2.31.0)
Requirement already satisfied: tomlkit>=0.7.2 in /usr/local/lib/python3.10/dist-packages (from munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (0.12.4)
Requirement already satisfied: PyYAML>=5.1 in /usr/lib/python3/dist-packages (from munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (5.4.1)
Requirement already satisfied: docutils in /usr/local/lib/python3.10/dist-packages (from python-daemon<3,>=2->vaping) (0.20.1)
Requirement already satisfied: lockfile>=0.10 in /usr/local/lib/python3.10/dist-packages (from python-daemon<3,>=2->vaping) (0.12.2)
Requirement already satisfied: setuptools in ./.local/lib/python3.10/site-packages (from python-daemon<3,>=2->vaping) (69.1.1)
Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests>=2.6->munge<2.0.0,>=1.2.0->munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (3.3.2)
Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests>=2.6->munge<2.0.0,>=1.2.0->munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (3.6)
Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests>=2.6->munge<2.0.0,>=1.2.0->munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (2.2.0)
Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests>=2.6->munge<2.0.0,>=1.2.0->munge[tomlkit,yaml]<2.0.0,>=1.2.0->vaping) (2024.2.2)
Downloading vaping-1.5.3-py3-none-any.whl (32 kB)
Installing collected packages: vaping
Successfully installed vaping-1.5.3
user@ubuntubox:~$ git clone https://github.com/20c/vaping
Cloning into 'vaping'...
remote: Enumerating objects: 2539, done.
remote: Counting objects: 100% (549/549), done.
remote: Compressing objects: 100% (208/208), done.
remote: Total 2539 (delta 368), reused 439 (delta 310), pack-reused 1990
Receiving objects: 100% (2539/2539), 3.69 MiB | 379.00 KiB/s, done.
Resolving deltas: 100% (1572/1572), done.
user@ubuntubox:~$ cd vaping/
user@ubuntubox:~/vaping$ vaping start --home=examples/standalone_dns/ --debug
Traceback (most recent call last):
  File "/home/user/.local/bin/vaping", line 5, in <module>
    from vaping.cli import cli
  File "/home/user/.local/lib/python3.10/site-packages/vaping/__init__.py", line 4, in <module>
    from vaping.config import Config  # noqa
  File "/home/user/.local/lib/python3.10/site-packages/vaping/config/__init__.py", line 5, in <module>
    from vaping.config.schema import *  # noqa
  File "/home/user/.local/lib/python3.10/site-packages/vaping/config/schema.py", line 1, in <module>
    import confu.exceptions
  File "/usr/local/lib/python3.10/dist-packages/confu/exceptions.py", line 5, in <module>
    import confu.schema
  File "/usr/local/lib/python3.10/dist-packages/confu/schema/__init__.py", line 1, in <module>
    from confu.schema.core import *  # noqa
  File "/usr/local/lib/python3.10/dist-packages/confu/schema/core.py", line 18, in <module>
    from confu.exceptions import ApplyDefaultError, ValidationError, ValidationWarning
ImportError: cannot import name 'ApplyDefaultError' from partially initialized module 'confu.exceptions' (most likely due to a circular import) (/usr/local/lib/python3.10/dist-packages/confu/exceptions.py)
```

**Comment by grizz** [comment]
Not abandoned, probably needs a little love tho. We'll take a look. :)

**Comment by marcogh** [comment]
Not a solution, but

```
pip install confu==1.8.0
```

fixed it

**Comment by smurfless1** [comment]
Confirmed updating to confu==1.8.0 fixed the docker image. Symptom: you can build the docker image, but on startup you get the error.

**Comment by doteater** [comment]
> Not abandoned, probably needs a little love tho. We'll take a look. :)

Thanks so much, I can start playing around with it now. I _might_ be in a position to make contributions in the future. Appreciate y'all.

**Comment by gizzygizmo** [comment]
Confirmed... adding
`pip install confu==1.8.0`
to Dockerfile fixed the build.

**Comment by grizz** [comment]
Thanks for all the info everyone.

It should be fixed in 1.5.4, please let us know if you're seeing any issues.
